### PR TITLE
Add support in LookupResources for caveats

### LINKED ISSUE #937: Add support for caveats in LookupResources API

### PR DESCRIPTION
**Pull request by josephschorr**
Fixes #937

In this initial implementation, any caveated paths are treated the same way an intersection or exclusion are treated, and a followup check is issued (with the proper context). There are almost certainly improvements we can make in the future to do proper book keeping of the actual caveat expressions and therefore just run the caveat expression without a full check, but as there are plans to optimize the checks as well, it remains to be seen whether that will be desirable

### CODE REVIEW

File: internal/dispatch/keys/computed_test.go [hunk 1]
```diff
@@ -115,7 +117,49 @@ func TestStableCacheKeys(t *testing.T) {
...
+							"bar": true,
+						})
+						return v
+					}(),
+				}, computeBothHashes)
+			},
+			"e3d8ffa3f7aa97e329",
+		},
+		{
+			"lookup resources with different context",
```
Comment (vroldanbet) [inline internal/dispatch/keys/computed_test.go:144]: can we add a test making sure the hash for `nil` context and `zero-value context` are the same? (hopefully 🤞🏻 )

File: internal/dispatch/keys/hasher_common.go [hunk 2]
```diff
@@ -63,3 +66,17 @@ type hashableString string
...
 }
+
+type hashableContext struct{ *structpb.Struct }
+
+func (hc hashableContext) AppendToHash(hasher hasherInterface) {
+	jsonBytes, err := hc.MarshalJSON()
+	if err != nil {
+		// If the struct could not be marshaled, use the string form of the struct.
+		logging.Err(err).Msg("got error marshaling context")
+		hasher.WriteString(hc.String())
```
Comment (vroldanbet) [inline internal/dispatch/keys/hasher_common.go:77]: If this were to happen, I fear it could somehow lead to a cache-key collision. Can we add some tests to exercise this? Shouldn't the request context have been validated in the API? if that's the case, maybe we should panic here as it would suggest an edge case?

File: internal/dispatch/keys/computed_test.go (hunk 1 above)
Comment (vroldanbet) [reply internal/dispatch/keys/computed_test.go:144]: If we are really going to move forward with serializing string representation of the context when serialization errors occur, can we at least add testing here?

Comment (vroldanbet) [reply internal/dispatch/keys/computed_test.go:144]: this would be a great spot for adding some fuzzing - please open an issue if you think that makes sense

File: internal/util/multimap.go [hunk 3]
```diff
@@ -1,15 +1,29 @@
...
 
+// MultiMap represents a map that can contain 1 or more values for each key.
 type MultiMap[T comparable, Q any] struct {
 	items map[T][]Q
 }
 
+// Clear clears all entries in the map.
+func (mm *MultiMap[T, Q]) Clear() {
+	mm.items = map[T][]Q{}
+}
```
Comment (vroldanbet) [inline internal/util/multimap.go:22]: Seems like dead code, can we remove?

File: internal/util/multimap.go [hunk 4]
```diff
@@ -1,15 +1,29 @@
 package util
 
+import (
+	"golang.org/x/exp/maps"
+)
+
+// NewMultiMap creates and returns a new multimap from keys of type T to values of type Q.
```
Comment (vroldanbet) [inline internal/util/multimap.go:7]:
```suggestion
// NewMultiMap creates and returns a new MultiMap from keys of type T to values of type Q.
```

File: internal/graph/computed/computecheck_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package graph
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/graph/computed/computecheck_test.go:1]: Added

File: internal/testfixtures/datastore.go [hunk 6]
```diff
@@ -214,6 +216,55 @@ func createTestCaveat(require *require.Assertions) []*core.CaveatDefinition {
 	}}
 }
 
+// DatastoreFromSchemaAndTestRelationships returns a validating datastore wrapping that specified,
+// loaded with the given scehma and relationships.
+func DatastoreFromSchemaAndTestRelationships(ds datastore.Datastore, schema string, relationships []*core.RelationTuple, require *require.Assertions) (datastore.Datastore, datastore.Revision) {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/testfixtures/datastore.go:221]: Tried, but it breaks a bunch of tests that are reliant upon the current ordering of things. I'd rather not block this PR on that. I've filed an issue: https://github.com/authzed/spicedb/issues/985

File: internal/services/v1/permissions.go [hunk 7]
```diff
@@ -358,26 +353,29 @@ func (ps *permissionServer) LookupResources(req *v1.LookupResourcesRequest, resp
...
-			)
+	for _, found := range lookupResp.ResolvedResources {
+		var partial *v1.PartialCaveatInfo
+		permissionship := v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION
+		if found.Permissionship == dispatch.ResolvedResource_CONDITIONALLY_HAS_PERMISSION {
+			permissionship = v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_CONDITIONAL_PERMISSION
+			partial = &v1.PartialCaveatInfo{
+				MissingRequiredContext: found.MissingRequiredContext,
+			}
 		}
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/services/v1/permissions.go:373]: I mean, it would in turn return the UNKNOWN outward which isn't great but also not *wrong*, per se. Thoughts?

File: internal/graph/parallelchecker.go [hunk 8]
```diff
@@ -2,9 +2,12 @@ package graph
...
 import (
 	"context"
-	"fmt"
 	"sync"
 
+	"golang.org/x/exp/maps"
+
+	"github.com/authzed/spicedb/internal/graph/computed"
+
 	"github.com/authzed/spicedb/internal/util"
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/graph/parallelchecker.go:11]: Darn auto-importer...

File: internal/graph/parallelchecker.go [hunk 9]
```diff
@@ -85,9 +88,17 @@ func (pc *parallelChecker) DepthRequired() uint32 {
...
 
-func (pc *parallelChecker) addResultsUnsafe(resourceID string) {
-	pc.foundResourceIDs.Add(resourceID)
-	if pc.foundResourceIDs.Len() >= int(pc.lookupRequest.Limit) {
+func (pc *parallelChecker) addResultsUnsafe(resolvedResource *v1.ResolvedResource) {
+	// If the result being added is conditional and we've already found a valid permission, skip.
+	if resolvedResource.Permissionship == v1.ResolvedResource_CONDITIONALLY_HAS_PERMISSION {
+		existing, ok := pc.foundResourceIDs[resolvedResource.ResourceId]
+		if ok && existing.Permissionship == v1.ResolvedResource_HAS_PERMISSION {
+			return
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/graph/parallelchecker.go:96]: Its hard to reproduce due to async ordering. I've added a unit test

File: internal/graph/parallelchecker.go [hunk 10]
```diff
@@ -106,7 +117,7 @@ func (pc *parallelChecker) QueueToCheck(resourceID string) {
 	queue := func() bool {
 		pc.mu.Lock()
 		defer pc.mu.Unlock()
-		if pc.foundResourceIDs.Len() >= int(pc.lookupRequest.Limit) {
+		if len(pc.foundResourceIDs) >= int(pc.lookupRequest.Limit) {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/graph/parallelchecker.go:120]: Consistency probably does, but I've added a unit test

File: internal/dispatch/graph/reachableresources_test.go [hunk 11]
```diff
@@ -9,12 +9,11 @@ import (
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/dispatch/graph/reachableresources_test.go:9]: Probably, but not in this PR

File: internal/dispatch/graph/reachableresources_test.go [hunk 12]
```diff
@@ -284,3 +286,314 @@ func BenchmarkReachableResources(b *testing.B) {
...
+				permission can_do_things = viewer & editor
+			}
+
+			caveat testcaveat(somecondition int) {
+				somecondition == 42
+			}
+			`,
+			[]*core.RelationTuple{
+				tuple.WithCaveat(tuple.MustParse("document:foo#viewer@user:tom"), "testcaveat"),
+				tuple.MustParse("document:foo#editor@user:tom"),
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/dispatch/graph/reachableresources_test.go:439]: Added, not sure how much value it gives

File: internal/dispatch/graph/reachableresources_test.go [hunk 13]
```diff
@@ -284,3 +286,314 @@ func BenchmarkReachableResources(b *testing.B) {
...
+				permission can_do_things = editor & viewer
+			}
+
+			caveat testcaveat(somecondition int) {
+				somecondition == 42
+			}
+			`,
+			[]*core.RelationTuple{
+				tuple.WithCaveat(tuple.MustParse("document:foo#viewer@user:tom"), "testcaveat"),
+				tuple.MustParse("document:foo#editor@user:tom"),
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/dispatch/graph/reachableresources_test.go:463]: We have non-caveated tests above

File: internal/dispatch/graph/reachableresources_test.go [hunk 14]
```diff
@@ -284,3 +286,314 @@ func BenchmarkReachableResources(b *testing.B) {
...
+				permission can_do_things = viewer - banned
+			}
+
+			caveat testcaveat(somecondition int) {
+				somecondition == 42
+			}
+			`,
+			[]*core.RelationTuple{
+				tuple.WithCaveat(tuple.MustParse("document:foo#viewer@user:tom"), "testcaveat"),
+				tuple.MustParse("document:foo#banned@user:tom"),
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/dispatch/graph/reachableresources_test.go:487]: We have non-caveated tests above

File: internal/dispatch/graph/reachableresources_test.go [hunk 15]
```diff
@@ -284,3 +286,314 @@ func BenchmarkReachableResources(b *testing.B) {
...
+			`,
+			[]*core.RelationTuple{
+				tuple.WithCaveat(tuple.MustParse("document:foo#viewer@user:tom"), "testcaveat"),
+				tuple.MustParse("document:foo#editor@user:tom"),
+			},
+			RR("document", "view"),
+			ONR("user", "tom", "..."),
+			[]reachableResource{
+				{"document:foo#view", true},
+				{"document:foo#view", false},
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/dispatch/graph/reachableresources_test.go:546]: Yes

File: internal/dispatch/keys/computed_test.go (hunk 1 above)
Comment (josephschorr) [reply internal/dispatch/keys/computed_test.go:144]: > can we add a test making sure the hash for `nil` context and `zero-value context` are the same? (hopefully 🤞🏻 )

Done

> If we are really going to move forward with serializing string representation of the context when serialization errors occur, can we at least add testing here?

Add testing for what?

> this would be a great spot for adding some fuzzing - please open an issue if you think that makes sense

Not sure we need fuzzing for cache key computation

File: internal/dispatch/keys/hasher_common.go (hunk 2 above)
Comment (josephschorr) [reply internal/dispatch/keys/hasher_common.go:77]: It won't, because the `String()` form of the struct is a proto text format: `fields:{key:"bar" value:{bool_value:true}} fields:{key:"foo" value:{number_value:1}}`

Comment (josephschorr) [reply internal/dispatch/keys/hasher_common.go:77]: Another option would be to panic

File: internal/util/multimap.go (hunk 3 above)
Comment (josephschorr) [reply internal/util/multimap.go:22]: Needed in the followup PR

**Comment by josephschorr** [comment]
Updated

File: internal/testfixtures/datastore.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/testfixtures/datastore.go:221]: Thanks for opening that!

File: internal/services/v1/permissions.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/services/v1/permissions.go:373]: would it? wouldn't the dispatching process cancel/timeout?

File: internal/graph/computed/computecheck_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vroldanbet) [reply internal/graph/computed/computecheck_test.go:1]: excellent!

File: internal/dispatch/keys/computed_test.go (hunk 1 above)
Comment (vroldanbet) [reply internal/dispatch/keys/computed_test.go:177]: >Add testing for what?

Add a test-case with a Struct that will err on `proto.Marshall` and have the hash key generated

File: internal/dispatch/keys/hasher_common.go (hunk 2 above)
Comment (vroldanbet) [reply internal/dispatch/keys/hasher_common.go]: yeah, panic is what I had in mind. Looks good 👍🏻

**Review by vroldanbet — CHANGES_REQUESTED** [review]
The `TestStableCacheKeys` tests serializing request content are flaking. Unfortunately, Google [deliberately added randomization](https://developers.google.com/protocol-buffers/docs/reference/go/faq#unstable-json) to prevent users from relying on it as a stable output.

>We make no promises about the long-term stability of Go's implementation of the [JSON format for protocol buffers](https://developers.google.com/protocol-buffers/docs/proto3#json). The specification only specifies what is valid JSON, but provides no specification for a canonical format for how a marshaler ought to exactly format a given message. To avoid giving the illusion that the output is stable, we deliberately introduce minor differences so that byte-for-byte comparisons are likely to fail.

>To gain some degree of output stability, we recommend running the output through a JSON formatted.

They basically say [_you are on your own_](https://developers.google.com/protocol-buffers/docs/reference/go/faq#hash):

>You need canonical serialization, where the marshaled output of a protocol buffer message is guaranteed to be stable over time. Unfortunately, no specification for canonical serialization exists at this time. You'll need to write your own or find a way to avoid needing one.

Here is an issue where they clarify [it won't be fixed](https://github.com/golang/protobuf/issues/1082), and here the offending commit in [`protobuf-go`](https://github.com/protocolbuffers/protobuf-go/commit/582ab3de426ef0758666e018b422dd20390f7f26)

**Review by vroldanbet — APPROVED** [review]
I've addressed the `protojson` unstable output in https://github.com/authzed/spicedb/pull/938/commits/06a5047f2c8835080f44ce57cf7a8b7c5c28d0f1.

LGTM 👍🏻

File: internal/services/v1/permissions.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (josephschorr) [reply internal/services/v1/permissions.go:373]: Why would it cancel?

File: internal/dispatch/keys/computed_test.go (hunk 1 above)
Comment (josephschorr) [reply internal/dispatch/keys/computed_test.go:177]: Not sure I can force that to error; I suspect that can only really happen if there is a memory issue. Anyway, I changed it to a panic

**Comment by josephschorr** [comment]
> The `TestStableCacheKeys` tests serializing request content are flaking. Unfortunately, Google [deliberately added randomization](https://developers.google.com/protocol-buffers/docs/reference/go/faq#unstable-json) to prevent users from relying on it as a stable output.

As discussed offline, I manually implemented sorting of keys and walking the struct directly

**Review by vroldanbet — APPROVED** [review]
